### ISSUE TITLE
Better aligning the way the BSSw site works with GitHub as a tool base

### ISSUE THREAD

**Issue by bernhold** [betterscientificsoftware/betterscientificsoftware.github.io#368]
This is intended to collect thoughts about how the way the BSSw.io site works, how content must be formatted, etc. could be better aligned with the GitHub ecosystem.  This is *not* intended for immediate action, but perhaps to set evolutionary goals.

* Use a git workflow, with branches, etc. to manage development, preview, and publication of content.
  - Right now, development is done in the GitHub repo, but preview and publication are triggered by metadata settings within each article
* The frontend should respond dynamically to new material in the Site folder, including *new* pages, *new* hierarchy, etc.
  - Right now, a lot of the content is fixed, at least at the level of specific pages being recognized by the frontend.
* Move metadata to YAML at the top of file instead of HTML comments at the end.
  - YAML is widely used in tools like Jekyll, etc.
  - YAML is presented in a reasonable way in the GitHub web-based file editing tools. It looks like a table, and more importantly, the metadata is actually *visible*.  The HTML comments are invisible because they are considered comments in MarkDown too
  - This would require modifying every piece of content on the site.  But it should be largely automatable.  It should also be possible to do it in two steps (having the YAML and HTML metadata coexist, at least long enough to validate that the frontend handles the YAML properly).
* Set repository permissions so that only the **EB** and **Editorial Assistants** have write access on the repo.  Everyone else operates by PR.
  - Currently, lots of people have write access
* Use hooks and CI tools with scripts to check the content in various ways
  - We already do a bit of this.  I'm not sure what.  I'm sure there's room for more.
  - Validating metadata
  - Spelling and grammar checking (needs to be a way to override if we do this)

**Comment by markcmiller86** [comment]
I'll add to this list...

- Put images *next to* the documents that reference them or, at the very least, in the *same* repo.
- Consider GIT LFS for binary content, like images and especially any video if we plan to support it.

**Comment by markcmiller86** [comment]
Forgot to mention another interesting issue regarding images...recording copyright info. In my blog articles on the AGC, I am using a lot of NASA images. In other articles, I've found images on line with appropriate usage rights and suggested them for use. 

We need a way to record, in the documents that use them, the copyright information for those images and maybe even links to where we got rights. In one case, I had an email exchange with a company.

**Comment by bernhold** [comment]
> * Consider GIT LFS for binary content, like images and especially any video if we plan to support it.

Another option to consider for videos (*if* we decide to go there) might be a YouTube channel.  We already have one for IDEAS (mostly hosting the HPC-BP webinar recordings).  YT has its own advantages and disadvantages, especially using it as an auxiliary to other sites.  I just don't know what is entailed in doing a proper job of hosting and presenting them ourselves.

**Comment by bernhold** [comment]
> Forgot to mention another interesting issue regarding images...recording copyright info. 

Agreed.  We need to re-think copyright for the whole site. And definitely for any media we use/host.

**Comment by markcmiller86** [comment]
> The frontend should respond dynamically to new material in the Site folder, including _new_ pages, _new_ hierarchy, etc.
> * Right now, a lot of the content is fixed, at least at the level of specific pages being recognized by the frontend.

I think if the site were Jekyll, Jekyll+Liquid+SASS basically serves this role. They are all fairly easy technologies to learn so in theory, we wouldn't need (a lot) of IT support.

**Comment by bernhold** [comment]
@betterscientificsoftware/bssw-editorial-board  I think we're well on our way to a more Github-centered process.  We're even using Jekyll and GH pages.

Have we done enough to warrant closing this?  Obviously we can create new tickets for issues that we do need to keep open.  I'm creating one about copyright.

**Comment by bartlettroscoe** [comment]
> I think we're well on our way to a more Github-centered process. We're even using Jekyll and GH pages.

@bernhold, in what respect?  For the bssw.io site itself, that is not using Jekyll or GitHub Pages, is it?  Only the EB Docs is using Jekyll and GitHub Pages.

**Comment by bernhold** [comment]
@bartlettroscoe yes, that's exactly correct.  It would be a far more significant change to introduce Jekyll for the site.  I think we have bigger fish to fry for the time being.

**Comment by bartlettroscoe** [comment]
> > > I think we're well on our way to a more Github-centered process. We're even using Jekyll and GH pages.
> >
> > For the bssw.io site itself, that is not using Jekyll or GitHub Pages, is it?
>
> yes, that's exactly correct. It would be a far more significant change to introduce Jekyll for the site.

@bernhold, then what is meant by "the BSSw site" in "Better aligning the way the BSSw site works with GitHub as a tool base"?  If "the BSSw site" means the production bssw.io site, when why mention "Jekyll and GH pages"?

If we are talking about the basic [GitHub Flow workflow](https://guides.github.com/introduction/flow/), then the new 'preview' branch workflow, moving the images so to the bssw.io GitHub Repo, where you create a single topic branch and PR branch, do all reviews and changes in that branch/PR and then merge/publish, then yes, what we are doing now is more compatible with how GitHub is intended to be used.  But given the scope of what is stated above, all of that is not done.  But what is done (or will be once we complete #738 next week) is #330 is complete (which was opened 2 months before this issue).

**Comment by bernhold** [comment]
Can we call this done?

**Comment by bartlettroscoe** [comment]
> Can we call this done?

@bernhold, as I mentioned [above](https://github.com/betterscientificsoftware/bssw.io/issues/368#issuecomment-850517337), everything listed in the scope of this [issue description](https://github.com/betterscientificsoftware/bssw.io/issues/368#issue-448977505) is not complete.  We should note what has been done and what has not been done by editing [that comment](https://github.com/betterscientificsoftware/bssw.io/issues/368#issue-448977505) and then close this issue.  I know some things have been done (like the git workflow with the preview branch) but I think you are in the best position to know how much of the rest of that list is completed because you deal with the contractor and the site generator.

**Comment by markcmiller86** [comment]
> Can we call this done?

Not sure what all being *done* entails here but I still routinely encounter situations where GFM doesn't render correctly on preview. I think we need this documented somewhere. Or, if there is a document for the markdown spec the backend is using, that might be ok too, as long as it isn't some homegrown thing.

**Comment by bernhold** [comment]
Sounds like maybe we should have a time-boxed discussion of this issue and decide whether we want to keep it open or close it.

**Comment by bartlettroscoe** [comment]
This issue is really an epic.  Many of the items listed above are stories and some of those have been completed.
